server/src/main/java/org/elasticsearch/index/mapper/TextFieldMapper.java

```diff
@@ -273,11 +329,27 @@ public void setFielddataMinSegmentSize(int fielddataMinSegmentSize) {
             this.fielddataMinSegmentSize = fielddataMinSegmentSize;
         }
 
+        public void setIndexPrefixes(int minChars, int maxChars) {
+            checkIfFrozen();
+            prefixAnalyzer = new PrefixWrappedAnalyzer(indexAnalyzer().analyzer(), minChars, maxChars);
+        }
+
         @Override
         public String typeName() {
             return CONTENT_TYPE;
         }
 
+        @Override
+        public Query prefixQuery(String value, MultiTermQuery.RewriteMethod method, QueryShardContext context) {
+            if (prefixAnalyzer == null || prefixAnalyzer.accept(value.length()) == false) {
+                return super.prefixQuery(value, method, context);
+            }
+            TermQuery q = new TermQuery(new Term(name() + "._prefix", indexedValueForSearch(value)));
```

Comment:
I don't think anything prevents a user from creating an explicit field with the same name?

Reply:
Not yet, no.  Do we have a way of reserving field names elsewhere?